### ISSUE TITLE
Question: Proposal data

### ISSUE THREAD

**Issue by YuhsinLiao1995** [forbole/bdjuno#139]
Hi, i got some issues when i was working on querying proposal related data for forbole x, they are listed below, could you help me figure them out? thanks!

1. `time_stamp` data of table of deposit details is not founded in bdjuno.
![截圖 2021-05-21 下午6 24 43](https://user-images.githubusercontent.com/61954949/119123178-e31f3780-ba61-11eb-978b-a876d656900b.png)

```
query Proposal($id: Int!) {
  proposal(where: {id: {_eq: $id }}) {
    proposal_deposits {
      amount
      depositor_address
      height
    }
  }
}
```

2.  every proposal has its own `status`, i checked it on bdjuno, they're only three proposal now, and their `status` are `PROPOSAL_STATUS_REJECTED`, `PROPOSAL_STATUS_INVALID`, `PROPOSAL_STATUS_REJECTED` respectively. can i assume that the `status` of voting, deposit proposal are `PROPOSAL_STATUS_VOTING_PERIOD` and `PROPOSAL_STATUS_DEPOSIT_PERIOD`?

3.  Every proposal need to show proposer's name and image, but now in `proposal` i can only get `proposer_address`, and then query `Account` to get the validator profile info. i'd like to know (1.) is it correct to display `validator_descriptions` as the info of the proposer? (2.) is it possible to add proposer data(name, avatar_url) under `Proposal` directly? it can save one query.
```
query Proposal {
  proposal {
    content
    description
    id
    title
    voting_end_time
    voting_start_time
    submit_time
    deposit_end_time
    proposal_type
    *proposer_address*
    status
  }
}
```
```
query Account($address: String!)  {
  account(where: {address: {_eq: $address}}) {
    address
    validator_infos {
      operator_address
      validator {
        validator_descriptions {
          avatar_url
          identity
          moniker
          validator_address
        }
      }
    }
  }
}
```

4. there is a diagram which display vote result detail, it need to show `name`, `image`, `voting power%`, `voting power`, `voting power override` of every vote, now i use `voter_addess` to query  `validator_descriptions` from `Account`, but i do not find `voting power%`, `voting power`, `voting power override` in `Account` now, (1.)is it possible to add this info under `proposal_vote`?  (2.)besides, is the column `option` represent the answer of the voter? ex: `VOTE_OPTION_YES` for yes?

![截圖 2021-05-21 下午7 18 04](https://user-images.githubusercontent.com/61954949/119129307-524c5a00-ba69-11eb-92dc-c404517d1aaa.png)
```
query VoteDetail($proposal_id: Int!) {
  proposal_vote(where: {proposal_id: {_eq: $proposal_id}}) {
    voter_address
    proposal_id
    option
  }
}
```

5. there is an absence tab in this table, how to query all the absence accounts?
<img width="592" alt="截圖 2021-05-24 下午12 11 18" src="https://user-images.githubusercontent.com/61954949/119295356-37f0c700-bc89-11eb-90a9-ebb06ac2aac9.png">

**Comment by rmontagnin** [comment]
#### 1. Proposal deposits
I've added the relation between a deposit `height` and the `block` table. You can now use
```graphql
query Proposal($id: Int!) {
  proposal(where: {id: {_eq: $id }}) {
    proposal_deposits {
      amount
      depositor_address
      block {
        timestamp
      }
    }
  }
}
```

**Comment by rmontagnin** [comment]
#### 2. Proposal statuses
The governance proposals can have the following statuses: 

- `PROPOSAL_STATUS_DEPOSIT_PERIOD` when they have been submitted and they are waiting for more deposits to be made; 
- `PROPOSAL_STATUS_INVALID` when the minimum deposit has not been reached and they have not passed to the voting phase; 
- `PROPOSAL_STATUS_VOTING_PERIOD` when the minimum deposit has been reached and they have passed to the voting phase; 
- `PROPOSAL_STATUS_PASSED` when the vote has ended and the proposal has been accepted;
- `PROPOSAL_STATUS_REJECTED`  when the vote has ended and the proposal was not accepted;  
- `PROPOSAL_STATUS_FAILED` when not enough votes have been sent and the proposal has failed.

**Comment by rmontagnin** [comment]
#### 3. Proposer details 
I've renamed the proposal field from `account` to `proposer` and now you can use the following query to get the details of the proposer immediately: 

```graphql
query Proposal {
  proposal {
    proposer {
      address
      validator_infos {
        validator {
          validator_descriptions {
            moniker
            avatar_url
          }
        }
      }
    }
  }
}
```

This should allow you to get the proposer image and description (that's represented by the `moniker`) immediately with one query. 

##### Note
Please note that a governance proposal can be created also by simple accounts that are not associated with any validator(s). In that case the `valiator_info` array will be empty and you will not be able to get the proposer image and description. If that happens, I think you want to show the default image as the icon and the `proposer.address` as the proposer description.

**Comment by rmontagnin** [comment]
#### 4. Proposal voting power 
I am going to move this into a separate issue, as it will require some discussion around it to make sure it's worth implementing this feature and I know how to do it properly. 

Edit: Moved into #140

**Comment by rmontagnin** [comment]
#### 5. Proposal absence 
To be honest, I have no idea how this feature was thought when it was being designed. I think you want to ask @kwunyeung, @ryuash or @Rooty2020 for this. In my opinion, it's going to be really hard to implement a true abscence inside proposals voting. Both validators and simple accounts can abstain from voting, but if the simple accounts abstain and their validator vote than should they be counted as absent? 

There are too many things that go on when dealing with absent voters that I would vouch against implementing it all together since it will become too difficult to do.

**Comment by ryuash** [comment]
@YuhsinLiao1995 @RiccardoM 
Absence only shows **validators**. But since we aren't keeping track of a validator's status history it might be a bit hard to implement. 

Pros: 
- we can see who is constantly not voting

Cons:
- we would need to keep a history of validator statuses by block

tldr: We decided it's best to put that feature on hold and focus on displaying the more direct data first.

**Comment by YuhsinLiao1995** [comment]
@RiccardoM 
could you add `validator_infos` under depositor_address too? thanks
```
query Proposal($id: Int!) {
  proposal(where: {id: {_eq: $id }}) {
    proposal_deposits {
      amount
      depositor_address
      block {
        timestamp
      }
    }
  }
}
```

**Comment by rmontagnin** [comment]
@YuhsinLiao1995 Renamed the `account` field into `depositor`. You can now use

```graphql
query Proposal($id: Int!) {
  proposal(where: {id: {_eq: $id}}) {
    proposal_deposits {
      amount
      depositor {
        validator_infos {
          ...
        }
      }
      block {
        timestamp
      }
    }
  }
}
```

**Comment by YuhsinLiao1995** [comment]
@RiccardoM thanks for helping!
could i possibly query `remaining deposit amount`?
![截圖 2021-05-25 下午8 17 22](https://user-images.githubusercontent.com/61954949/119496795-6f996500-bd96-11eb-9948-bfa042a48301.png)

**Comment by rmontagnin** [comment]
@YuhsinLiao1995 You need to: 

1. Query the governance deposit params
2. Sum all the deposits that have been made to a proposal
3. The remaining amount will be `remaining = deposit_params.min_deposit - sum(deposits)`

You can get the params and the deposits with the following query: 

```graphql
query RemainingDeposit($proposalId: Int) {
  gov_params {
    deposit_params
  }
  proposal_deposit(where: {proposal_id: {_eq: $proposalId}}) {
    amount
  }
}
```

You need to sum the amounts by hand though (make sure you only sum the amounts with the same denom)

**Comment by YuhsinLiao1995** [comment]
@RiccardoM i got it, thanks!
